src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,221 @@
+        return;
+    }
+
+    auto dns_cb = [=](std::string service, std::string hostname) {
+        return [=](Error err, Var<dns::Message> message) {
+            if (!!err) {
+                logger->info("fb_messenger: dns error for %s, %s",
+                    service.c_str(), hostname.c_str());
+            } else {
+                for (auto answer : message->answers) {
+                    if ((answer.ipv4 != "") || (answer.hostname != "")) {
+                        std::string asn_p = options.get("geoip_asn_path", std::string{});
+                        auto geoip = GeoipCache::thread_local_instance()->get(asn_p);
+                        ErrorOr<std::string> asn = geoip->resolve_asn(answer.ipv4);
+                        if (asn && asn.as_value() != "AS0") {
+                            logger->info("%s ipv4: %s, %s",
+                                hostname.c_str(), answer.ipv4.c_str(),
+                                asn.as_value().c_str());
+                            // if consistent, add to list for tcp connect later
+                            if (asn.as_value() == "AS32934") {
+                                (*fb_service_ips)[service].push_back(answer.ipv4);
+                            } else {
+                                (*entry)["facebook_dns_blocking"] = true;
+                            }
+                        }
+                    }
+                }
+                // if any consistent IPs, consistent = true
+                (*entry)["facebook_" + service + "_dns_consistent"] =
+                    !(*fb_service_ips)[service].empty();
+            }
+            *names_tested += 1;
+            if (names_count == *names_tested) {
+                cb(NoError(), entry, *fb_service_ips, options, reactor, logger);
+                return;
+            }
+        };
+    };
+
+    for (auto const& service_and_hostname : fb_service_hostnames) {
+        std::string service = service_and_hostname.first;
+        std::string hostname = service_and_hostname.second;
+        templates::dns_query(entry, "A", "IN", hostname, "",
+                             dns_cb(service, hostname), options, reactor,
+                             logger);
+    }
+    return;
+}
+
+static void tcp_many(Error error,
+                     Var<Entry> entry,
+                     std::map<std::string, std::vector<std::string>> fb_service_ips,
+                     Settings options,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Var<Entry>> cb
+                    ) {
+    logger->info("starting tcp_many");
+    if (error) {
+        cb(entry);
+        return;
+    }
+
+    // if we find any blocked TCP later, switch this to true
+    (*entry)["facebook_dns_blocking"] = false;
+
+    int ips_count = 0;
+    for (auto const& service_and_ips : fb_service_ips) {
+        // skipping stun for now
```

Comment:
Why do we skip `stun`?

Reply:
Because ooni-probe does :) and I think because it uses UDP? *me reads* I guess it uses both, but we'd only be able to check the TCP bit?

Reply:
Ah, okay: can you document that we're doing what ooni-probe does in the comment?